libraries/postgres/src/Database.js

```diff
@@ -0,0 +1,292 @@
+        }
+      }
+
+      for (let exp of expectedPrivs) {
+        if (!currentPrivs.has(exp)) {
+          issues.push(`missing database user grant: ${exp}`);
+        }
+      }
+
+      if (issues.length > 0) {
+        throw new Error(`Database privileges are not configured as expected:\n${issues.join('\n')}`);
+      }
+    });
+  }
+
+  async _doUpgrade({version, showProgress, usernamePrefix}) {
+    await this._withClient('admin', async client => {
+      await client.query('begin');
+      if (version.version === 1) {
+        await client.query('create table tcversion as select 0 as version');
+      }
+
+      // check the version and lock it to prevent other things from changing it
+      const res = await client.query('select version from tcversion for update');
+      if (res.rowCount !== 1 || res.rows[0].version !== version.version - 1) {
+        throw Error('Multiple DB upgrades running simultaneously');
+      }
+      showProgress('..running migration script');
+      const migrationScript = version.migrationScript
+        .replace('$db_user_prefix$', usernamePrefix);
+      await client.query(`DO ${dollarQuote(migrationScript)}`);
+      showProgress('..defining methods');
+      for (let [methodName, { args, body, returns}] of Object.entries(version.methods)) {
+        await client.query(`create or replace function
+        "${methodName}"(${args})
+        returns ${returns}
+        as ${dollarQuote(body)}
+        language plpgsql`);
+      }
+      showProgress('..updating version');
+      await client.query('update tcversion set version = $1', [version.version]);
+      showProgress('..committing transaction');
+      await client.query('commit');
+    });
+  }
+
+  /**
+   * Private constructor (use Database.setup and Database.upgrade instead)
+   */
+  constructor({urlsByMode, statementTimeout}) {
+    const makePool = dbUrl => {
+      const pool = new Pool({connectionString: dbUrl});
+      // ignore errors from *idle* connections
+      pool.on('error', client => {});
+      pool.on('connect', async client => {
+        if (statementTimeout) {
+          await client.query(`set statement_timeout = ${statementTimeout}`);
+        }
+        // unconditionally apply a timeout for idle transactions. we should
+        // never be idle in a transaction (well, for more than few ms beteween
+        // statements), so this is set quite low.  Holding a transaction open
+        // can make locks pile up and also prevent vacuuming of tables, both
+        // leading to performance issues.
+        await client.query('set idle_in_transaction_session_timeout = 200');
+      });
+      return pool;
+    };
+
+    this.pools = {};
+    for (let mode of Object.keys(urlsByMode)) {
+      this.pools[mode] = makePool(urlsByMode[mode]);
+    }
+
+    this.fns = {};
+  }
+
+  /**
+   * Run cb with a client.
+   *
+   * This is for INTERNAL USE ONLY.  All external access to the DB should be
```

Comment:
maybe "LIBRARY USE ONLY" I interpreted this as internal to TC. I know that is nonsensical but still